### ISSUE TITLE
Allow for registering mocks with ApplicationContextRunner

### ISSUE THREAD

**Issue by davidkarlsen** [spring-projects/spring-boot#13718]
It would be nice if one could register mock beans with the `ApplicationContextRunner` in some easy way.
`@MockBeans` in the test does not seem to be picked up.
Sometimes the auto-configuration interacts with other beans, and being able to easily register them for each test would be beneficial.

**Comment by wilkinsona** [comment]
Neither `@MockBean` nor `@SpyBean` will work as they're intended for use with `@SpringBootTest`. `ApplicationContextRunner` is intended for use in a "plain" test using a standard runner.

Have you considered using a normal `@Configuration` class that provides the necessary mock beans? That's what we do in our own tests such as [this one](https://github.com/spring-projects/spring-framework/blob/master/spring-core/src/main/java/org/springframework/core/type/classreading/AnnotationAttributesReadingVisitor.java) and we haven't felt a need for anything more.

**Comment by philwebb** [comment]
Perhaps something like `new ApplicationContextRunner().withMockBeans(SomeBean.class).withUserConfiguration(MyConfiguration.class)` might be useful.

**Comment by wilkinsona** [comment]
That would help for basic scenarios where you do not need to configure any behaviour on the mock, but I don't think it would help much when you do. You also lose the ability to specify things like the name of the bean or its order. I guess `withMockBeans` could take something that richer that captures all of that, but it feels like quite a lot of work for not much benefit.

**Comment by davidkarlsen** [comment]
withMockBeans looks just like what I need. The behaviour will be given per test anyway ( in the run() method - at the start) - but the bean needs to be there. I went the way via a separate @ConfigurationClass given via withUserConfig - but it's tedious and verbose.

**Comment by spring-issuemaster** [comment]
If you would like us to look at this issue, please provide the requested information. If the information is not provided within the next 7 days this issue will be closed.

**Comment by davidkarlsen** [comment]
My code looks like this:
`
public class WsContextResolverAutoConfigurationTest
    extends AbstractIntegrationTest
{

    private final ApplicationContextRunner contextRunner = new ApplicationContextRunner()
            .withConfiguration(
                    AutoConfigurations.of(
                            WsContextResolverAutoConfiguration.class,
                            OracleEndToEndMetricsAutoConfiguration.class));

    @Configuration
    public static class TestConfig {

        @Bean
        public WebServiceContext webServiceContext() {
            return new WebServiceContextImpl();
        }

        /* shold work...:
        @Bean
        public OracleDataSource oracleEndToEndMetricDataSourceWrapper() {
            return Mockito.mock(OracleDataSource.class);
        }
        */

        @Bean
        public OracleEndToEndMetricDataSourceWrapper pu() {
            return Mockito.mock(OracleEndToEndMetricDataSourceWrapper.class);
        }
    }

    @Test
    public void testNotActive() {
        contextRunner.run( c -> Assertions.assertThat(c).doesNotHaveBean(WsContextResolver.class));
    }

    @Test
    public void testActive() {
        contextRunner
                .withPropertyValues(
                        "evry.fs.jfr.cxf.utils.oracle-metrics.headerClass=com.edb.finance.common.cxf.generated.wsheaders.wscommon_v22.EDBHeaderType",
                        "evry.fs.jfr.cxf.utils.oracle-metrics.headerPropertyPathToOrgId=test")
                .withUserConfiguration(TestConfig.class)
                .run( c -> Assertions.assertThat(c)
                        .hasSingleBean(WsContextResolver.class)
                        .getBean(WsContextResolver.class)
                        .hasFieldOrPropertyWithValue("headerClass", EDBHeaderType.class)
                        .hasFieldOrPropertyWithValue("headerPropertyPathToOrgId", "test"));
    }
}
`

basically I need a mock of a specific type to get the autoconfig accepted

**Comment by wilkinsona** [comment]
Thanks. In a comment above you said “The behaviour will be given per test anyway ( in the run())”. I can’t see any configuration of the mock’s behaviour in the example above. Can you please clarify?

**Comment by davidkarlsen** [comment]
ah - that's in another test, here you go:
```
 @Configuration
    @Import(AbstractDataSourceConfiguration.class)
    public static class TestConfig {

        @Bean
        public JdbcTemplate getJdbcTemplate( DataSource dataSource ) {
            return new JdbcTemplate( dataSource );
        }

        // https://github.com/spring-projects/spring-boot/issues/13718
        @Bean
        public OracleEndToEndMetricDataSourceWrapper.ContextResolver contextResolver() {
            OracleEndToEndMetricDataSourceWrapper.ContextResolver contextResolver = Mockito.mock(OracleEndToEndMetricDataSourceWrapper.ContextResolver.class);

            return contextResolver;
        }
    }

  @Test
    public void testDbOperation() {
        contextRunner.withUserConfiguration(TestConfig.class).run( c -> {
            JdbcTemplate jdbcTemplate = c.getBean(JdbcTemplate.class);
            Assertions.assertThat(jdbcTemplate.getDataSource()).isInstanceOf(OracleEndToEndMetricDataSourceWrapper.class);
            OracleEndToEndMetricDataSourceWrapper.ContextResolver contextResolver = c.getBean(OracleEndToEndMetricDataSourceWrapper.ContextResolver.class);

                Mockito.when(contextResolver.resolve()).thenReturn( new OracleEndToEndMetricDataSourceWrapper.Context( "someModule",
                    "someAction",
                    "someClientid",
                    "someExecutionContextId",
                    Short.valueOf( "1" ),
                    RandomStringUtils.randomAscii( 30 ) ) );

            jdbcTemplate.queryForObject( "select 1 from dual", Integer.class );
            Mockito.verify(contextResolver).resolve();

        });
    }
```

**Comment by philwebb** [comment]
I guess not all tests will be fortunate enough to use mocks like that. In many causes I imagine the initialization will need to be done as part of the `@Bean` method.

Given the limited usefulness of `withMockBeans` alone, and the added complexity that would be required if we were to offer a way to setup the mock, I think we should not rush to add this. Using the `@Bean` method to create the mock is a little inconvenient, but not too bad.
